provenance.jq

```diff
@@ -0,0 +1,80 @@
+# input: "build" object (with "buildId" top level key)
+# output: array of image tags
+def tags:
+  .source.arches[].tags[],
+  .source.arches[].archTags[],
+  .build.img
+;
+
+# input: "build" object (with "buildId" top level key)
+# output: purl platform query string
+def platform_string:
+  .source.arches[].platformString | gsub("/"; "%2F")
+;
+
+# input: "tags" object with image digest and platform arguments
+# output: json object for in-toto provenance subject field
+def subjects($platform; $digest):
+  {
+      "name": ("pkg:docker/" + . + "?platform=" + $platform),
+      "digest": {
+        "sha256": $digest
+      }
+  }
+;
+
+# input: GITHUB context argument
+# output: json object for in-toto provenance external parameters field
+def github_external_parameters($context):
+($context.workflow_ref | gsub( $context.repository + "/"; "")) as $workflowPathRef |
+{
+  inputs: $context.event.inputs,
+  workflow: {
+    ref: ($workflowPathRef | split("@")[1]),
+    repository: ($context.server_url + "/" + $context.repository),
+    path: ($workflowPathRef | split("@")[0]),
+    digest: {sha256: $context.workflow_sha}
+  }
+}
+;
+
+# input: GITHUB context argument
+# output: json object for in-toto provenance internal parameters field
+def github_internal_parameters($context):
+{
+  github: {
+    event_name: $context.event_name,
+    repository_id: $context.repository_id,
+    repository_owner_id: $context.repository_owner_id,
+  }
+}
+;
+
+# input: "tags" object with platform, image digest and GITHUB context arguments
+# output: json object for in-toto provenance statement
+def github_actions_provenance($platform; $digest; $context):
```

Comment:
It was originally collapsed down and everything took `builds.json` (see https://github.com/docker-library/meta-scripts/commit/6fcc5db4fa9e386d169afb28b8c3d714a372741f) but we split it apart so that the build job would output `platform` and `tags` such that the `sign` job wouldn't need to checkout the large `builds.json` because it only needs build tags and platform data. 

We could switch this back but would want @LaurentGoderre to weigh in too since we made this change together.

Reply:
> maybe there's a subset of that full context document we could pass in/around?

This is a good point, we could just pass in all of the individual fields from the `github` context that we need for provenance. We aren't using any sensitive fields for this but they do exist in the `github` context. Specifically, `github.token`. I am not very worried about the usage here but if you want to we can selectively pass in what we need, maybe make a JSON object from the fields (but not using JQ?) and pass it in that way.

Reply:
here is a list of what we are currently using:
```
server_url
repository
repository_id
repository_owner_id
ref
workflow_ref
workflow_sha
event_name
event.inputs
run_id
run_attempt
```

Reply:
so, alternatively we could set the context var like so:
```bash
context=$(jq -n \
  --arg server_url "${{ github.server_url }}" \
  --arg repository "${{ github.repository }}" \
  --arg repository_id "${{ github.repository_id }}" \
  --arg repository_owner_id "${{ github.repository_owner_id }}" \
  --arg ref "${{ github.ref }}" \
  --arg workflow_ref "${{ github.workflow_ref }}" \
  --arg workflow_sha "${{ github.workflow_sha }}" \
  --arg event_name "${{ github.event_name }}" \
  --arg event_inputs "${{ github.event_inputs }}" \
  --arg run_id "${{ github.run_id }}" \
  --arg run_attempt "${{ github.run_attempt }}" \
  '{
    server_url: $server_url,
    repository: $repository,
    repository_id: $repository_id,
    repository_owner_id: $repository_owner_id,
    ref: $ref,
    workflow_ref: $workflow_ref,
    workflow_sha: $workflow_sha,
    event_name: $event_name,
    event_inputs: $event_inputs,
    run_id: $run_id,
    run_attempt: $run_attempt
  }')
```

It may be worth it to make sure the token isn't mishandled. Thoughts on adding that?

Reply:
I prefer the simplicity of just passing the entire `github` context as a `--jsonarg` since I think the risk of mishandling the `github.token` is relatively low (especially since it is masked in the output). But willing to go either way or see if you have any other ideas for it.

Reply:
> It was originally collapsed down and everything took `builds.json` (see [6fcc5db](https://github.com/docker-library/meta-scripts/commit/6fcc5db4fa9e386d169afb28b8c3d714a372741f)) but we split it apart so that the build job would output `platform` and `tags` such that the `sign` job wouldn't need to checkout the large `builds.json` because it only needs build tags and platform data from that file (and we are already consuming it in the build job).

Hmm yeah, I see your point, although that does also make it slightly harder to add more interesting information from it to this document later, if we wanted to. :thinking:

Do you think there would be much overhead in doing something like providing the full "build" object as an output (in JSON format) so that we could pass along _just_ the object from the large file that we're working on?  That would give us something somewhere in the middle so that this is still consistent with most of our other interfaces, easier to run/test locally, and still avoiding the full clone in the provenance generation step. :eyes:

> I prefer the simplicity of just passing the entire `github` context as a `--jsonarg` since I think the risk of mishandling the `github.token` is relatively low (especially since it is masked in the output). But willing to go either way or see if you have any other ideas for it.

Yeah, it's really not ideal either way -- I with GitHub would add more functionality for filtering/munging objects in their expressions so it didn't have to be _so_ manual (or provide a second copy of the context/a function that filters out any sensitive data and makes it safer for raw consumption/use). :confused:

So I guess passing the full context is probably the lesser of two evils and we'll continue to be careful about what we do with it -- we should probably also include a warning similar to that in the upstream documentation to help remind us that the context object is generally sensitive when used in full. :eyes:

Also, as a general point, I think we should avoid using `${{ ... }}` entirely _directly_ in `run:` blocks to help practice our discipline of separation and avoid accidents (https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable), so that should go in an environment variable, which can then be used in `jq` directly via something like `(env.GITHUB_CONTEXT | fromjson) as $github | ...`, which _also_ helps avoid command-line length limits (and instead pushes them into environment size limits, but those limits are much higher). :smile:

Reply:
Using environment variable works well, fixed that. We are working on refactoring/testing the build outputs now